### DISCUSSION TITLE
Addbookmark and IBookmarkresult

### DISCUSSION THREAD

**Discussion by makattar** [elsa-workflows/elsa-core#1248] [Q&A]
I am referring https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-blocking-activities .Addbookmarkprovider method and IBookmarkresut not found as error message.And i am using elsa 2.0.

**Comment by sfmskywalker** [comment]
Hi @makattar ,

Indeed that guide was written against Elsa 2.1.
For Elsa 2.0, I believe you can return the bookmark object directly.

**Reply by makattar** [reply]
Hi Sir, 
But I am having trouble also when collecting workflow context because It requires third argument named Trigger(type - IBookmark) 
**new CollectWorkflowsContext(nameof(FileReceived), new FileReceivedBookmark());** 
What should i add on third argument?
I tried to add null , But doesnt execute activity.

**Reply by sfmskywalker** [reply]
Right. This extra parameter is likely going to be removed since it’s a bit redundant, but until then you should pass in the same value as for the bookmark parameter.

**Comment by makattar** [comment]
Hi sir.I tried to pass second parameter as null and third parameter named Trigger as new object bookmark.And now it's worked.
Again one thing sir, Now it's triggering all workflows where this activity is used.
So the problem is how can I filter which workflow to execute.
Like how Http endpoint activity filters I want the same functionality.

**Reply by sfmskywalker** [reply]
> I tried to pass second parameter as null and third parameter named Trigger as new object bookmark

That should indeed work for new workflows, but the second parameter (taking a bookmark object) is important when trying to resume existing workflow instances.

> Now it's triggering all workflows where this activity is used.

Do you mean when you have multiple workflow definitions that use your activity as the starting activity? If so, then it makes sense. What you probably want then is to be able to further specify any additional conditions under which the activity should trigger.

For example, the **Signal Received** activity only triggers when:

- A bookmark exists for that activity type, and:
- That bookmark contains the same **signal** name that was configured for the activity.

Similarly, when dealing with e.g. a **File Received** activity, perhaps you want to configure **from where** the file was received, or perhaps the file extension name, or some other aspect that allows you to setup different workflows for the same File Received event but differentiated by certain aspects of the file being received.

If you mean that even existing workflow instances that are **blocked** on your activity all execute when e.g. a file is received, then not only do you need to consider on what aspects of the file you want to differentiate, but you also might need to consider how to **correlate** your workflow instances.

If you describe your exact use case in a bit more detail, I might be able to help you come up with a good strategy.

**Reply by makattar** [reply]
Okay sir..

My use case scenario where I wants to use -

So I have created MyActivity there...MyActivity is same as like Filereceived activity.
MyActivity has an activity property named entity. I will create workflows with MyActivity and set Entity property as Student or Employee.When I will dispatch workflow same as used in Filereceivedinvoker ,I will pass entity value there.Then only workflows which is having MyActivity and entity property set as passed entity value should invoked .


One more question -
As per you said second parameter is important .So should I have to write like
new CollectWorkflowsContext(nameof(FileReceived), new FileReceivedBookmark(),
new FileRecievedBookmark());
I tried but it's not invoking workflows.

**Reply by sfmskywalker** [reply]
Perfect. In that case, you should add a property to your `FileReceivedBookmark` called `Entity`. And of course update your bookmark provider to read the selected entity for the activity being indexed and store it in the bookmark.

For Elsa 2.0, it is important that you provide two bookmark objects as you did: one for resuming existing workflow instances, and one for starting new workflows.

In Elsa 2.1, you can only provide one bookmark that is used for both resuming and starting workflows.

You mention that it's not invoking workflows. It's hard to say why without seeing the full code in action.

**Reply by makattar** [reply]
Ok sir Thank you.

**Reply by makattar** [reply]
Sir can you explain me with FileRecived activity as per documentation just change is that there is dropdown for supportedExtension.(With code example)

**Reply by sfmskywalker** [reply]
Good idea, I will update the documentation guide to show how to extend the activity with a `SupportedExtension` property displayed as a dropdown and how to use it when building a bookmark.

**Reply by sfmskywalker** [reply]
I just looked at the documentation and it already has a property called `SupportedFileExtensions`, so I assume you want to know what it would take to change that to a singular `SupportedFileExtension` displayed as a dropdown.

To do that, do the following:

1. Rename the property to `SupportFileExtension`.
2. Change its type to `string?`.
3. Change its `UIHint` to `ActivityInputUIHints.Dropdown`.
4. Remove `DefaultSyntax`.
5. Remove `DefaultValue`.
6. Change `SupportedSyntaxes` to `new[] {SyntaxNames.JavaScript, SyntaxNames.Liquid})]`.
7. Update `FileReceivedBookmarkProvider` to return a single bookmark for the specified extension.

The updated property would look like this:

```csharp
[ActivityInput(
    Hint = "Specify a file extensions to filter. Leave empty to allow any file extension.",
    UIHint = ActivityInputUIHints.Dropdown, 
    SupportedSyntaxes = new[] {SyntaxNames.JavaScript, SyntaxNames.Liquid})]
public string? SupportedFileExtension { get; set; }
```

And the updated bookmark provider would look like this:

```csharp
public class FileReceivedBookmarkProvider : BookmarkProvider<FileReceivedBookmark, FileReceived>
{
    public override async ValueTask<IEnumerable<BookmarkResult>> GetBookmarksAsync(BookmarkProviderContext<FileReceived> context, CancellationToken cancellationToken)
    {
        var supportedExtension = (await context.ReadActivityPropertyAsync<FileReceived, string?>(x => x.SupportedFileExtension, cancellationToken));

        return Result(new FileReceivedBookmark(supportedExtension));
    }
}
```

I haven't tested the code, so let me know if you run into any issues.
